### PR TITLE
fix identifying CREATE_FUNCTION for complex postgres functions

### PR DESCRIPTION
**Pull request by MasterOdin**
Previous PR for this (#17) only used the base examples out of the user manual and failed to account that these statements can be quite complex, including multiple `END` tokens as well as semi-colons.

This PR addresses that shortcoming with a more robust implementation that should work for postgres, taking advantage of the fact that the `LANGUAGE` token is mandatory for created functions to use that as a signifier of the end of the part of the statement that can contain semi-colons. An alternative implementation would be to track "block openers" (`BEGIN`, `IF`, `LOOP`, etc.), but I'm not sure how useful or fully cross dialect such that I think just sticking to the specific postgres method might be better for the time being.

### CODE REVIEW

File: test/identifier/single-statement.spec.js [hunk 1]
```diff
@@ -97,12 +97,39 @@ describe('identifier', function () {
...
+             USING employee.staff_id, total_bonus, sales_total, start_date, end_date;
+        END LOOP;
+        DELETE FROM sales WHERE created_at >= start_date
+               AND created_at < end_date;
+        RETURN QUERY SELECT name, bonus, period FROM sales_summary
+                     LEFT JOIN staff on sales_summary.staff_id = staff.staff_id;
+       RETURN;
+      END;
+      $$
+      LANGUAGE plpgsql;`;
```
Comment (rathboma) [inline test/identifier/single-statement.spec.js:126]: I think unfortunately it's valid for `LANGUAGE` to be before the `BEGIN` block, which would throw out the logic of your parser.

For example this query from https://stackoverflow.com/questions/8918755/postgresql-function-with-a-loop
switching `LANGUAGE` to before `AS` it is still valid:

```sql
CREATE OR REPLACE FUNCTION f_grp_prod(text)
  RETURNS TABLE (
    name text
  , result1 double precision
  , result2 double precision)
LANGUAGE plpgsql STABLE
AS
$BODY$
DECLARE
    r      mytable%ROWTYPE;
    _round integer;
BEGIN
    -- init vars
    name    := $1;
    result2 := 1;       -- abuse result2 as temp var for convenience

FOR r IN
    SELECT *
    FROM   mytable m
    WHERE  m.name = name
    ORDER  BY m.round
LOOP
    IF r.round <> _round THEN   -- save result1 before 2nd round
        result1 := result2;
        result2 := 1;
    END IF;

    result2 := result2 * (1 - r.val/100);
    _round  := r.round;
END LOOP;

RETURN NEXT;

END;
$BODY$;

```

See for example - https://www.postgresqltutorial.com/postgresql-create-function/

**Review by rathboma — CHANGES_REQUESTED** [review]
Don't think this will work :-/

I think this requires figuring out which statements have an `END` - Loops, Conditionals, case?

File: test/identifier/single-statement.spec.js (hunk 1 above)
Comment (MasterOdin) [reply test/identifier/single-statement.spec.js:126]: You are absolutely right. I've included this function as another test case, and modified the code to account for this, adjusting it to fit my latter thought on how to accomplish this (keeping track of block openers).

**Review by rathboma — CHANGES_REQUESTED** [review]
CASE in particular requires an `END` across all databases.

File: src/parser.js [hunk 2]
```diff
@@ -22,7 +22,9 @@ const EXECUTION_TYPES = {
   UNKNOWN: 'UNKNOWN',
 };
 
-const dialectsWithEnds = ['sqlite', 'mssql'];
+const dialectsWithOpenBlocks = ['psql'];
+const blockOpeners = ['BEGIN', 'IF', 'LOOP', 'CASE'];
```
[the remark this answers is not in the document]
Comment (MasterOdin) [reply src/parser.js:26]: You are correct. I think for now, let's target:

* `CASE` is used across DBs.
* `BEGIN` is used for postgres, sql server, and mysql.
* `LOOP` and `IF` are used in postgres and mysql.

[the remark this answers is not in the document]
Comment (MasterOdin) [reply src/parser.js]: I've updated the code to include the above, and added test cases for it.

**Comment by MasterOdin** [comment]
Alright, I think for now we've got a good number of the edge cases sussed out that I would say merge this and cut a new release, and can do further PRs for edge cases as they come up?

**Review by rathboma — APPROVED** [review]
Awesome work! A++

:shipit:
